src/client/rag.ts

```diff
@@ -0,0 +1,415 @@
+import { v4 as uuidv4 } from "uuid";
+import { App, AuthAppConfig } from "./app";
+import { Workflow } from "./workflow";
+import * as fs from "fs";
+import yaml from "js-yaml";
+import {
+  JavaScriptValue,
+  Struct,
+} from "google-protobuf/google/protobuf/struct_pb";
+import { Model } from "./model";
+import { User } from "./user";
+import { MAX_UPLOAD_BATCH_SIZE } from "../constants/rag";
+import { UserError } from "../errors";
+import { AuthConfig } from "../utils/types";
+import { ClarifaiAppUrl, ClarifaiUrl, ClarifaiUrlHelper } from "../urls/helper";
+import {
+  ModelVersion,
+  OutputInfo,
+} from "clarifai-nodejs-grpc/proto/clarifai/api/resources_pb";
+import { validateWorkflow } from "../workflows/validate";
+import {
+  Message,
+  convertMessagesToStr,
+  formatAssistantMessage,
+  loadDocuments,
+  splitDocument,
+} from "../rag/utils";
+import { Input } from "./input";
+import compact from "lodash/compact";
+
+const DEFAULT_RAG_PROMPT_TEMPLATE =
+  "Context information is below:\n{data.hits}\nGiven the context information and not prior knowledge, answer the query.\nQuery: {data.text.raw}\nAnswer: ";
+
+type UrlAuthConfig = Omit<AuthConfig, "userId" | "appId"> & {
+  appId?: undefined;
+  userId?: undefined;
+};
+
+type RAGConfig =
+  | {
+      workflowUrl: ClarifaiUrl;
+      workflow?: undefined;
+      authConfig?: UrlAuthConfig;
+    }
+  | {
+      workflowUrl?: undefined;
+      workflow: Workflow;
+      authConfig?: AuthConfig;
+    };
+
+type workflowSchema = ReturnType<typeof validateWorkflow>;
+
+export class RAG {
+  private authConfig: AuthConfig;
+
+  public promptWorkflow: Workflow;
+
+  public app: App;
+
+  constructor({ workflowUrl, workflow, authConfig }: RAGConfig) {
```

Comment:
I have narrowed down the types with typeguards but I'm still leaving the runtime checks in place since we are expecting javascript users to also try this sdk who will not benefit from Typescript linters alone